### ISSUE TITLE
igl::sparse_voxel_grid outputting duplicate vertex positions

### ISSUE THREAD

**Issue by alecjacobson** [libigl/libigl#1690]
`igl::sparse_voxel_grid` intends to output a _**unique**_ list of vertex positions in `CV` so that indices in `CI` build the cubes of a sparse grid. For certain inputs, vertex positions will be (exactly) duplicated. Here's rather minimal example:

```cpp
#include <igl/sparse_voxel_grid.h>
#include <iostream>

int main(int argc, char * argv[])
{
  const auto f = [&](const Eigen::RowVector3d & x)->double
  {
    return x.norm() - 1.0;
  };
  Eigen::RowVector3d p0(0,1.0,0);
  Eigen::MatrixXd GV;
  Eigen::VectorXd Gf;
  Eigen::Matrix<int,Eigen::Dynamic,8> GI;
  igl::sparse_voxel_grid(p0,f,1,1024,Gf,GV,GI);
  std::cout<<"GV.row(58) "<<GV.row(58)<<std::endl;
  std::cout<<"GV.row(75) "<<GV.row(75)<<std::endl;
  std::cout<<":-"<<(GV.row(58).isApprox(GV.row(75),0)?"(":")")<<std::endl;
}
```

This prints:
```
GV.row(58) 1.5 1.5 1.5
GV.row(75) 1.5 1.5 1.5
:-(
```

The code in `sparse_voxel_grid.cpp` is hard to parse. I can't tell if the bug is in the hash function or in the bit masking or somewhere else in the logic.

**Comment by alecjacobson** [comment]
@fwilliams not sure if you still remember this code enough to help track down this bug.

**Comment by sgsellan** [comment]
I have encountered this in the past, and solved it by looping over all (26) neighbors and not just the (6) face neighbors. The relevant changes in the code are 


```
const std::array<Eigen::RowVector3i, 26> neighbors = {
      bx, -bx, by, -by, bz, -bz,
      by-bz, -by+bz, // 1-2 4-7
      bx+by, -bx-by, // 0-1 7-6
      by+bz, -by-bz, // 0-3 6-5
      by-bx, -by+bx, // 2-3 5-4
      bx-bz, -bx+bz, // 1-5 3-7
      bx+bz, -bx-bz, // 0-4 2-6
      -bx+by+bz, bx-by-bz, // 3 5
      bx+by+bz, -bx-by-bz, // 0 6
      bx+by-bz, -bx-by+bz, //1 7
      -bx+by-bz, bx-by+bz, // 2 4,
}
constexpr std::array<uint8_t, 26> zv = {
      (1 << 0) | (1 << 1) | (1 << 4) | (1 << 5),
      (1 << 2) | (1 << 3) | (1 << 6) | (1 << 7),
      (1 << 0) | (1 << 1) | (1 << 2) | (1 << 3),
      (1 << 4) | (1 << 5) | (1 << 6) | (1 << 7),
      (1 << 0) | (1 << 3) | (1 << 4) | (1 << 7),
      (1 << 1) | (1 << 2) | (1 << 5) | (1 << 6),
      (1 << 1) | (1 << 2),
      (1 << 4) | (1 << 7),
      (1 << 0) | (1 << 1),
      (1 << 6) | (1 << 7),
      (1 << 0) | (1 << 3),
      (1 << 5) | (1 << 6),
      (1 << 2) | (1 << 3),
      (1 << 4) | (1 << 5),
      (1 << 1) | (1 << 5),
      (1 << 3) | (1 << 7),
      (1 << 0) | (1 << 4),
      (1 << 2) | (1 << 6),
      (1 << 3), (1 << 5), // diagonals
      (1 << 0), (1 << 6),
      (1 << 1), (1 << 7),
      (1 << 2), (1 << 4),
}
constexpr std::array<std::array<int, 4>, 26> zvv {{
      {{0, 1, 4, 5}}, {{3, 2, 7, 6}}, {{0, 1, 2, 3}},
      {{4, 5, 6, 7}}, {{0, 3, 4, 7}}, {{1, 2, 5, 6}},
      {{-1,-1,1,2}}, {{-1,-1,4,7}}, {{-1,-1,0,1}},{{-1,-1,7,6}},
      {{-1,-1,0,3}}, {{-1,-1,5,6}}, {{-1,-1,2,3}}, {{-1,-1,5,4}},
      {{-1,-1,1,5}}, {{-1,-1,3,7}}, {{-1,-1,0,4}}, {{-1,-1,2,6}},
      {{-1,-1,-1,3}}, {{-1,-1,-1,5}}, {{-1,-1,-1,0}}, {{-1,-1,-1,6}},
      {{-1,-1,-1,1}}, {{-1,-1,-1,7}}, {{-1,-1,-1,2}}, {{-1,-1,-1,4}}}
```

Then, all loops must be changed to ignore any `-1` entry

**Comment by alecjacobson** [comment]
A hot fix (assuming the bug is _perfectly_ duplicating vertices):

```
  igl::sparse_voxel_grid(p0,f,1,1024,Gf,GV,GI);
  % sparse_voxel_grid is buggy and duplicating vertices.
    {
      Eigen::VectorXi I,_;
      igl::remove_duplicate_vertices(decltype(GV)(GV),decltype(GI)(GI),0,GV,I,_,GI);
      igl::slice(decltype(Gf)(Gf),I,1,Gf);
    }
```

**Comment by fwilliams** [comment]
It's been quite some time since I wrote this, but I can take a look. @alecjacobson do you know if it is it perfectly duplicating vertices or just duplicating vertices that are adjacent to multiple grid cells?

**Comment by alecjacobson** [comment]
I might not fully understand the distinction, but the vertices are coming
out exactly the same if that helps.

>

**Comment by fwilliams** [comment]
The distinction I was trying to refer to was if the duplication falls into one of the following cases:
 - _Case 1:_ If the sparse voxel grid is defined as `GV, GI`, then for any index `i <= GV.rows()`, is there always exactly one other index `j != i` such that `GV.row(i) == GV.row(j)`. i.e. the vertices are perfectly duplicated twice. 
 - _Case 2:_ The vertices are duplicated more than twice but always the same number of times?
 - _Case 3:_ The vertices are duplicated but a different amount of times per vertex. 

Also are you seeing the same bug for all inputs you've tried?

I looked over the code and I think [this line](https://github.com/libigl/libigl/blob/780a5fbea5596bfce9809f7c5845315f14ac8ad2/include/igl/sparse_voxel_grid.cpp#L111) is suspect. The bit shift array `zv` is a mask per cube face corresponding to the indices of vertices adjacent to that face (which we use to build a bitmask of vertices which are already added). `zvv` are the integer indices of cube vertices adjacent to a face. The `n % 2 == 0 n + 1 : n - 1` is weird and I don't remember what I was doing there. I need to draw the picture of what this code is doing and hopefully I'll see the issue. I'll keep you posted!

**Comment by fwilliams** [comment]
Not sure if this the cause of the issue here, but there is a bug in this code where vertices will be duplicated if a new cube gets added and there are more than one neighbor of that cube already in the sparse grid. 

This could be fixed by maintaining a hash table of vertices (in integer coordinates like grid centers), but doing extra hashing could be slow. For medium size inputs, it might make sense to simply remove the duplication checking logic and do the `igl::remove_duplicate_vertices` fix you proposed at the end of the function for now (this would at least make the code much simpler).

**Comment by alecjacobson** [comment]
Thanks. Did you see silvia's fix? It seems to work by checking a larger
neighborhood.

**Comment by fwilliams** [comment]
Ah yes, I just reread the code Silvia posted. The first time I looked through this thread, the array of bitshifts scared me away from reading it too closely, but it makes sense now after re-reading the original code. Looking at all 26 neighbors should indeed fix the issue I described above! @sgsellan after applying your fix, do you see duplicated vertices on any of the inputs you've tried? If not, then we could merge it as a fix.

**Comment by alecjacobson** [comment]
I'm marking this as fixed via https://github.com/libigl/libigl/commit/44f2e1d8bee59dc0e1710634717ac6856c870ef1 . I don't have enough recollection to determine if checking all 26 is overkill, but seems it's at worst a performance bug we could try to fine tune in the future. @sgsellan it'd be reassuring if you commented that you're using this fixed sparse_voxel_grid function successfully now.
